api/apps/api/src/modules/access-control/projects-acl/project-acl.service.ts

```diff
@@ -0,0 +1,82 @@
+import { InjectRepository } from '@nestjs/typeorm';
+import { Injectable } from '@nestjs/common';
+import { Repository } from 'typeorm';
+import { intersection } from 'lodash';
+
+import { UsersProjectsApiEntity } from '@marxan-api/modules/projects/control-level/users-projects.api.entity';
+import { Roles } from '@marxan-api/modules/users/role.api.entity';
+
+import { Permit, ProjectAccessControl } from './project-access-control';
+
+/**
+ * Debt: neither UsersProjectsApiEntity should belong to projects
+ * nor the Roles should belong to users
+ */
+@Injectable()
+export class ProjectAclService implements ProjectAccessControl {
+  private readonly canCreateScenarioRoles = [Roles.project_owner];
+  private readonly canEditScenarioRoles = [Roles.project_owner];
+  private readonly canViewSolutionRoles = [Roles.project_owner];
+  private readonly canPublishProjectRoles = [Roles.project_owner];
+  private readonly canViewProjectRoles = [
+    Roles.project_owner,
+    Roles.project_contributor,
+    Roles.project_viewer,
+  ];
+
+  private async getRolesWithinProject(
+    userId: string,
+    projectId: string,
+  ): Promise<Array<string>> {
+    const rolesToCheck = (
+      await this.roles.find({
+        where: {
+          projectId,
+          userId,
+        },
+        select: ['roleName'],
+      })
+    ).flatMap((role) => role.roleName);
+    return rolesToCheck;
+  }
+
+  constructor(
+    @InjectRepository(UsersProjectsApiEntity)
+    private readonly roles: Repository<UsersProjectsApiEntity>,
+  ) {}
+
+  // TODO: this will be changed in the following release of user requirements.
+  // For now, anyone should be able to create projects, regardless of having roles or not.
+  // In the future project creation will be limited to organization contributors, so this logic will be moved to the access control module
+  async canCreateProject(_userId: string, _projectId: string): Promise<Permit> {
+    return true;
+  }
+
+  async canViewProject(userId: string, projectId: string): Promise<Permit> {
+    const roles = await this.getRolesWithinProject(userId, projectId);
+
+    return intersection(roles, this.canViewProjectRoles).length > 0;
```

Comment:
this is very personal opinion/preference - I too do expose some raw "computation" snippets in my own code, but these lines repeated in each can*Project() method to me seem to be screaming asking to be wrapped in a more expressively-named function that we can use in each can<Action><Resource>() method across the app.

```suggestion
    return doesUserHaveRole(
        await this.getRolesWithinProject(userId, projectId),
        this.canViewProjectRoles) ? Allow : Deny;
```

What I mean, repeating the mechanics of checking the cardinality of set intersections sounds tedious. If we can identify a good enough abstraction for the time being, I'd abstract it.

But again, it's also partly a matter of preference.

(Also, it can be abstracted in several ways - above I have suggested a light touch, but the function could also be something like `doUserRolesAllow(roles: Roles[], wantedRoles: Role[]): Promise<Permit>`, for example)